src/widgets/dialogs/ColorPickerDialog.cpp

```diff
@@ -0,0 +1,365 @@
+{
+    auto cpPanel = creator.setLayoutType<QHBoxLayout>();
+
+    /*
+     * For some reason, LayoutCreator::emplace didn't work for these.
+     * (Or maybe I was too dense to make it work.)
+     * After trying to debug for 4 hours or so, I gave up and settled
+     * for this solution.
+     */
+    auto *colorPicker = new QColorPicker(this);
+    this->ui_.picker.colorPicker = colorPicker;
+
+    auto *luminancePicker = new QColorLuminancePicker(this);
+    this->ui_.picker.luminancePicker = luminancePicker;
+
+    cpPanel.append(colorPicker);
+    cpPanel.append(luminancePicker);
+
+    QObject::connect(colorPicker, SIGNAL(newCol(int, int)), luminancePicker,
+                     SLOT(setCol(int, int)));
+
+    QObject::connect(
+        luminancePicker, &QColorLuminancePicker::newHsv,
+        [=](int h, int s, int v) {
+            int alpha = this->ui_.picker.spinBoxes[SpinBox::ALPHA]->value();
+            this->selectColor(QColor::fromHsv(h, s, v, alpha), true);
+        });
+}
+
+void ColorPickerDialog::initSpinBoxes(LayoutCreator<QWidget> &creator)
+{
+    auto spinBoxes = creator.setLayoutType<QGridLayout>();
+
+    auto *red = this->ui_.picker.spinBoxes[SpinBox::RED] =
+        new QColSpinBox(this);
+    auto *green = this->ui_.picker.spinBoxes[SpinBox::GREEN] =
+        new QColSpinBox(this);
+    auto *blue = this->ui_.picker.spinBoxes[SpinBox::BLUE] =
+        new QColSpinBox(this);
+    auto *alpha = this->ui_.picker.spinBoxes[SpinBox::ALPHA] =
+        new QColSpinBox(this);
+
+    // We need pointers to these for theme changes
+    auto *redLbl = this->ui_.picker.spinBoxLabels[SpinBox::RED] =
+        new QLabel("Red:");
+    auto *greenLbl = this->ui_.picker.spinBoxLabels[SpinBox::GREEN] =
+        new QLabel("Green:");
+    auto *blueLbl = this->ui_.picker.spinBoxLabels[SpinBox::BLUE] =
+        new QLabel("Blue:");
+    auto *alphaLbl = this->ui_.picker.spinBoxLabels[SpinBox::ALPHA] =
+        new QLabel("Alpha:");
+
+    spinBoxes->addWidget(redLbl, 0, 0);
+    spinBoxes->addWidget(red, 0, 1);
+
+    spinBoxes->addWidget(greenLbl, 1, 0);
+    spinBoxes->addWidget(green, 1, 1);
+
+    spinBoxes->addWidget(blueLbl, 2, 0);
+    spinBoxes->addWidget(blue, 2, 1);
+
+    spinBoxes->addWidget(alphaLbl, 3, 0);
+    spinBoxes->addWidget(alpha, 3, 1);
+
+    for (size_t i = 0; i < SpinBox::END; ++i)
+    {
+        QObject::connect(
+            this->ui_.picker.spinBoxes[i],
+            QOverload<int>::of(&QSpinBox::valueChanged), [=](int value) {
+                this->selectColor(QColor(red->value(), green->value(),
+                                         blue->value(), alpha->value()),
+                                  false);
+            });
+    }
+}
+
+void ColorPickerDialog::initHtmlColor(LayoutCreator<QWidget> &creator)
+{
+    auto html = creator.setLayoutType<QGridLayout>();
+
+    // Copied from Qt source for QColorShower
+    QRegularExpression regExp(
```

Comment:
Will this be compiled every time the function is called?

Reply:
Good catch – made `static` in f4d82f2.